lib/ipam/ipam.go

```diff
@@ -299,10 +301,44 @@ func (c ipamClient) autoAssign(ctx context.Context, num int, handleID *string, a
 		logCtx = logCtx.WithField("handle", *handleID)
 	}
 	logCtx.Info("Looking up existing affinities for host")
-	affBlocks, err := c.blockReaderWriter.getAffineBlocks(ctx, host, version, pools)
+	affBlocks, affBlocksToRelease, err := c.blockReaderWriter.getAffineBlocks(ctx, host, version, pools)
 	if err != nil {
 		return nil, err
 	}
+
+	// Release any blocks still affine to this host but no longer selected by the larger ip pool.
+	for _, block := range affBlocksToRelease {
+		// Determine the pool for each block.
+		pool := c.blockReaderWriter.getPoolForIP(cnet.IP{block.IP}, allPools)
+		if pool == nil {
+			logCtx.WithFields(log.Fields{"pool": pool, "block": block}).Warn("No pool found for block, skipping")
+			continue
+		}
+
+		// Determine if the pool selects the current node, refusing to release this particular block affinity if so.
+		blockSelectsNode, err := pool.SelectsNode(*v3n)
+		if err != nil {
+			logCtx.WithError(err).WithField("pool", pool).Error("Failed to determine if node matches pool, skipping")
+			continue
+		}
+		if blockSelectsNode {
+			logCtx.Debug("Block's pool still selects node, refusing to remove affinity")
+			continue
+		}
+
+		// If we made it this far, relase the block affinity.
+		if err = c.blockReaderWriter.releaseBlockAffinity(ctx, host, block, true); err != nil {
+			if _, ok := err.(errBlockClaimConflict); ok {
+				// Not claimed by this host - ignore.
+			} else if _, ok := err.(cerrors.ErrorResourceDoesNotExist); ok {
+				// Block does not exist - ignore.
+			} else {
+				return nil, err
```

Comment:
Just had a thought, but I wonder if we need to retry here on conflict errors, or if it's OK to simply ignore errors.

Think of this scenario:
- We list all the blocks we think we should release
- We go to release a block, but in the meantime someone else has allocated an IP address from that block.
- We fail with an update conflict error

In that scenario, we no longer want to release the block since it isn't empty any more. If we retry the `releaseBlockAffinity`, it will query the node, determine that it isn't empty, and carry-on as expected. The current logic won't retry, and instead will return an error, failing the allocation. WDYT?